src/peer_select.cc

```diff
@@ -457,57 +580,100 @@ PeerSelector::selectMore()
 
     debugs(44, 3, request->method << ' ' << request->url.host());
 
-    /** If we don't know whether DIRECT is permitted ... */
-    if (direct == DIRECT_UNKNOWN) {
-        if (always_direct == ACCESS_DUNNO) {
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (always_direct to be checked)");
-            /** check always_direct; */
-            ACLFilledChecklist *ch = new ACLFilledChecklist(Config.accessList.AlwaysDirect, request, NULL);
-            ch->al = al;
-            acl_checklist = ch;
-            acl_checklist->nonBlockingCheck(CheckAlwaysDirectDone, this);
-            return;
-        } else if (never_direct == ACCESS_DUNNO) {
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (never_direct to be checked)");
-            /** check never_direct; */
-            ACLFilledChecklist *ch = new ACLFilledChecklist(Config.accessList.NeverDirect, request, NULL);
-            ch->al = al;
-            acl_checklist = ch;
-            acl_checklist->nonBlockingCheck(CheckNeverDirectDone, this);
-            return;
-        } else if (request->flags.noDirect) {
-            /** if we are accelerating, direct is not an option. */
-            direct = DIRECT_NO;
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (forced non-direct)");
-        } else if (request->flags.loopDetected) {
-            /** if we are in a forwarding-loop, direct is not an option. */
-            direct = DIRECT_YES;
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (forwarding loop detected)");
-        } else if (checkNetdbDirect()) {
-            direct = DIRECT_YES;
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (checkNetdbDirect)");
-        } else {
-            direct = DIRECT_MAYBE;
-            debugs(44, 3, "direct = " << DirectStr[direct] << " (default)");
+    while (!currentServer && selectionState != DoFinished) {
+        switch (selectionState) {
+        case DoCheckDirect:
+            checkDirect();
+            if (selectionState == DoCheckDirect)
+                return; // wait nonblocking acls to finish
+            break;
+        case DoPinned:
+            assert(!entry || entry->ping_status == PING_NONE);
+            selectPinned();
+            break;
+        case DoSelectSomeNeighbors:
+            selectSomeNeighbor();
+            break;
+        case DoStartPing:
+            assert(entry);
+            startIcpPing();
+            break;
+        case DoContinuePing:
+            continueIcpPing();
+            if (selectionState == DoContinuePing)
+                return; // wait for pinging results
+            break;
+        case DoFinalizePing:
+            finalizeIcpPing();
+            break;
+        case DoFinal:
+            finalSelections();
+            break;
+        default:
+            // not reached
```

Comment:
Ok, but I still wonder whether the `selectionState == DoFinished` condition is possible within the loop. How can it happen?

Reply:
It can not happen.
But why to assert on this? Even in the case in the future for a reason this is can happen, it will not cause any problem the sendNextPeer will handle correctly this case.
If you think it is important I will add the assertion.

Reply:
Ok, leaving as is.